File: client/src/main/java/io/xol/chunkstories/client/net/packets/PacketInitializeRemoteWorld.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.client.net.packets;

import io.xol.chunkstories.api.content.OnlineContentTranslator;
import io.xol.chunkstories.api.net.PacketReceptionContext;
import io.xol.chunkstories.api.net.PacketSender;
import io.xol.chunkstories.client.net.ClientPacketsEncoderDecoder;
import io.xol.chunkstories.net.packets.PacketSendWorldInfo;
import io.xol.chunkstories.world.WorldInfoUtilKt;

import java.io.DataInputStream;
import java.io.IOException;

public class PacketInitializeRemoteWorld extends PacketSendWorldInfo {

	public void process(PacketSender sender, DataInputStream in, PacketReceptionContext processor) throws IOException {
		String initializationString = in.readUTF();

		worldInfo = WorldInfoUtilKt.deserializeWorldInfo(initializationString);

		if (processor instanceof ClientPacketsEncoderDecoder) {
			processor.logger().info("Received World initialization packet");
			ClientPacketsEncoderDecoder cpp = (ClientPacketsEncoderDecoder) processor;

			OnlineContentTranslator contentTranslator = cpp.getContentTranslator();
			if (contentTranslator == null) {
				processor.logger().error("Can't initialize a world without a ContentTranslator initialized first!");
				return;
			}

			//TODO remake this mechanism but make it actually any good
			/*
			IngameClientRemoteHost client = (IngameClientRemoteHost) cpp.getContext();
			Fence fence = client.getGameWindow().queueSynchronousTask(new Runnable() {
				@Override
				public void run() {
					WorldClientRemote world;
					try {
						world = new WorldClientRemote(client, worldInfo, contentTranslator, cpp.getConnection());
						client.changeWorld(world);

						cpp.getConnection().handleSystemRequest("world/ok");
					} catch (WorldLoadingException e) {
						client.exitToMainMenu(e.getMessage());
					}
				}
			});

			fence.traverse();*/
		}
	}
}

File: client/src/main/java/xyz/chunkstories/client/net/packets/PacketInitializeRemoteWorld.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.client.net.packets

import xyz.chunkstories.api.Engine
import xyz.chunkstories.api.server.UserConnection
import xyz.chunkstories.net.packets.PacketSendWorldInfo
import xyz.chunkstories.world.deserializeWorldInfo
import java.io.DataInputStream
import java.io.IOException

class PacketInitializeRemoteWorld(engine: Engine) : PacketSendWorldInfo(engine) {

    @Throws(IOException::class)
    override fun receive(dis: DataInputStream, user: UserConnection?) {
        val initializationString = dis.readUTF()

        worldInfo = deserializeWorldInfo(initializationString)

        /*if (processor is ClientPacketsEncoderDecoder) {
            processor.logger().info("Received World initialization packet")

            val contentTranslator = processor.contentTranslator
            if (contentTranslator == null) {
                processor.logger().error("Can't initialize a world without a ContentTranslator initialized first!")
                return
            }

            val sequence = processor.connection.connectionSequence
            val client = sequence.client

            val ingame = IngameClientRemoteHost(client, processor.connection) {
                WorldClientRemote(it as IngameClientRemoteHost, worldInfo, processor.contentTranslator as AbstractContentTranslator, processor.connection)
            }
            ingame.world.startLogic()
            client.ingame = ingame

            sequence.worldSemaphore.release()

            //TODO("remake this mechanism but make it actually any good")
            //throw new RuntimeException("TODO");
            //val client = processor.client

                    /*val client = processor.context as IngameClientRemoteHost
                    (client.gameWindow as GLFWWindow).mainThread {
                        val world: WorldClientRemote
                        try {
                            world = WorldClientRemote(client, worldInfo, contentTranslator as AbstractContentTranslator, processor.connection)
                            client.gui.topLayer = IngameLayer()
                            //(client as ClientImplementation).changeWorld(world)

                            processor.connection.handleSystemRequest("world/ok")
                        } catch (e: WorldLoadingException) {
                            client.exitToMainMenu(e.message!!)
                        }
                    }*/
        } else {
            error("shit")
        }*/
        TODO()
    }
}
